bookkeeper-server/src/main/java/org/apache/bookkeeper/meta/MetadataDrivers.java

```diff
@@ -167,17 +167,28 @@ private static void loadInitialBookieDrivers() {
      */
     public static void registerClientDriver(String metadataBackendScheme,
                                             Class<? extends MetadataClientDriver> driver) {
+        registerClientDriver(metadataBackendScheme, driver, false);
+    }
+
+    @VisibleForTesting
+    public static void registerClientDriver(String metadataBackendScheme,
+                                            Class<? extends MetadataClientDriver> driver,
+                                            boolean allowOverride) {
         if (!initialized) {
             initialize();
         }
 
         String scheme = metadataBackendScheme.toLowerCase();
         MetadataClientDriverInfo oldDriverInfo = clientDrivers.get(scheme);
-        if (null != oldDriverInfo) {
+        if (null != oldDriverInfo && !allowOverride) {
             return;
         }
         MetadataClientDriverInfo newDriverInfo = new MetadataClientDriverInfo(driver);
-        oldDriverInfo = clientDrivers.putIfAbsent(scheme, newDriverInfo);
+        if (allowOverride) {
```

Comment:
It'd be better to handle all the special case branching based on a boolean flag in one place.
```java
if (null != oldDriverInfo) {
    if (allowOverride) {
        log.debug("Metadata client driver for {} is already there.", scheme); 
    } else {
        return;
    }
}
clientDrivers.put(scheme, newDriverInfo);
```

Reply:
this changes the logging behavior though. the logging message is added after `put`, that is to handle concurrent puts. the change you are proposing here doesn't do that.

Reply:
Ah, true. Still, it would be nice to check allow override only once, to so the read doesn't have to keep so much state in mind while reading. How about.
```
MetadataClientDriverInfo oldDriverInfo = clientDrivers.computeIfAbsent(
        scheme, (k) -> new MetadataClientDriverInfo(driver));
if (null != oldDriverInfo) {
    log.debug("Metadata client driver for {} is already there.", scheme);
    if (allowOverride) {
        log.debug("Overriding client driver for {}", scheme);
        clientDrivers.put(scheme, new MetadataClientDriverInfo(driver));
    }
}
```

Reply:
not sure this is better than current implementation. would prefer stick to current implementation unless there is a huge concern here.

Reply:
My problem with this is that allowOverride is brought into the default flow of the method. It's even the first branch from the if on L183. It means multiple returns from the method. Logging only occurs if there is a race. The existence in the map is checked twice. And what should be the non-default path is intertwined with the default.
None of these are huge in themselves, but they combine to make me dislike the current flow a lot. Another alternative, which would be a least deal with some of the intertwining...
```
MetadataClientDriverInfo oldDriverInfo = clientDrivers.get(scheme);
if (null != oldDriverInfo && !allowOverride) {
    return;
}
MetadataClientDriverInfo newDriverInfo = new MetadataClientDriverInfo(driver);
oldDriverInfo = clientDrivers.putIfAbsent(scheme, newDriverInfo);
if (null != oldDriverInfo) {
    log.debug("Metadata client driver for {} is already there.", scheme);
    if (allowOverride) {
        log.debug("Overriding client driver for {}", scheme);
        clientDrivers.put(scheme, newDriverInfo);
    }
}
```